tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +182,160 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
     return buffer_page_mapping;
 }
 
-Buffer::Buffer(const Buffer &other) :
-    device_(other.device_),
-    size_(other.size_),
-    page_size_(other.page_size_),
-    buffer_type_(other.buffer_type_),
-    buffer_layout_(other.buffer_layout_),
-    shard_parameters_(other.shard_parameters_),
-    bottom_up_(other.bottom_up_),
-    buffer_page_mapping_(other.buffer_page_mapping_),
-    allocate_(other.allocate_) {
-    if (this->allocate_) {
-        this->allocate();
+Buffer::Buffer(
+    Device *device,
+    DeviceAddr size,
+    DeviceAddr page_size,
+    const BufferType buffer_type,
+    const TensorMemoryLayout buffer_layout,
+    const std::optional<ShardSpecBuffer>& shard_parameters,
+    const std::optional<bool> bottom_up) :
+    device_(device),
+    size_(size),
+    page_size_(page_size),
+    buffer_type_(buffer_type),
+    buffer_layout_(buffer_layout),
+    shard_parameters_(shard_parameters),
+    bottom_up_(bottom_up),
+    buffer_page_mapping_(nullptr) {
+    TT_FATAL(this->device_ != nullptr && this->device_->allocator_ != nullptr, "Device and allocator need to not be null.");
+
+    if (size != 0) {
+        validate_buffer_size_and_page_size(size, page_size, buffer_type, buffer_layout, shard_parameters);
+    }
+}
+
+std::shared_ptr<Buffer> Buffer::create(
+    Device *device,
+    DeviceAddr size,
+    DeviceAddr page_size,
+    const BufferType buffer_type,
+    const TensorMemoryLayout buffer_layout,
+    const std::optional<ShardSpecBuffer>& shard_parameters,
+    const std::optional<bool> bottom_up) {
+    auto* bufferPtr = new Buffer(device, size, page_size, buffer_type, buffer_layout, shard_parameters, bottom_up);
+    auto buffer = std::shared_ptr<Buffer>(bufferPtr, deleter);
+    buffer->weak_self = buffer;
+
+    if (buffer->size_ == 0) {
+        buffer->allocation_status_.store(AllocationStatus::ALLOCATED, std::memory_order::relaxed);
+        return buffer;
     }
+
+    buffer->device_->push_work([buffer] {
+        bool bottom_up = buffer->bottom_up_.value_or(buffer->is_dram());
+        buffer->address_ = detail::AllocateBuffer(buffer.get(), bottom_up);
+        detail::BUFFER_MAP.insert({buffer->device_->id(), buffer->address_}, buffer.get());
+
+        buffer->allocation_status_.store(AllocationStatus::ALLOCATED, std::memory_order::release);
+        buffer->allocation_status_.notify_all();
+    });
+
+    return buffer;
+}
+
+void Buffer::deallocate() {
+    deallocation_requested_.store(true, std::memory_order::relaxed);
```

Comment:
Id use acquire here.

Reply:
Doesn't really matter either way, we're not using this flag for any synchronization of any sort